certlint_test.go

```diff
@@ -46,12 +50,36 @@ accZ5e1iOtNIBTVUEaQ0YxNwhZqnQO4rgBFCnEpNH6TB62hXUD6u/oP/WB48Wmek
 KPgaoWHMo5sOFQnw5A==
 -----END CERTIFICATE-----`
 
+func TestMain(m *testing.M) {
```

Comment:
Is there a reason you're invoking `exec.Command` to build the test data? I might be overlooking something but `go test ./...` should build as required no?

Same Q: re the `go clean` invocation. I'm not sure I understand why its needed. Maybe a comment would help explain?

Reply:
Unfortunately, no. It doesn't seem to automatically build a CLI binary for tests without this trickery in TestMain.

However, I am extremely new to writing tests for go (CLI tool or otherwise), so I would be elated if I were wrong and could delete the whole TestMain trickery. If it were actually building a binary, would it land in `~/.cache/go-build/`, or somewhere else? If it landed in the cache, how would I be able to predict the name in order to execute the CLI utility for testing?

Perhaps a point to be made here is that test code would be much more useful as further unit tests rather than CLI utility tests (essentially integration tests). Or, CLI utility tests would be much better as travis tests?

Reply:
Ahhh! I understand better now. I hadn't really clued in that this is more of an integration test and needs to run the project binaries.

I think one way to approach this case might be to have the `.travis.yml` `before_script` stage (or equiv) run `go install` on the project packages. The `certlint` binary should end up in `$GOPATH/bin` and I _think_ that's on the travis `$PATH` for Go projects by default. You would be able to remove the code you have to build the `certlint` binary and `certlint_test.go` would be just test code and nothing related to build/install.

I might be sending you on a wild goose chase based on vague theories :-) I'll try to actually test this when it's not Friday night. Def. don't consider my input here a blocker.

Reply:
Not use if adding a build to the `before_script` in `.travis.yml` would really help here, this would prevent you from running the test locally before committing. Unless you do the build manually.

Reply:
I did give this a short test in https://github.com/cpu/certlint/commit/beeb7241cde419179c5678e1be6a3cfd3d67922b and using a `go install` worked OK and seemed cleaner than the cross-concern of building/installing from within the test code.

RE: local testing, it seems reasonable to assume that someone who wants to run the integration tests should have the tool installed (e.g. has run `go install ./...` in the repo). You are correct that if the integration test didn't also build & install the binaries you wouldn't be able to run it without doing so yourself.